### PR TITLE
[zero] support all-gather overlap

### PR DESCRIPTION
**Pull request by ver217**
## 📌 Checklist before creating the PR

- [ ] I have created an issue for this PR for traceability
- [ ] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [ ] I have added relevant tags if possible for us to better distinguish different PRs
- [ ] I have installed pre-commit: `pip install pre-commit && pre-commit install`

## 🚨 Issue number

> Link this PR to your issue with words like fixed to automatically close the linked issue upon merge
>
> e.g. `fixed #1234`, `closed #1234`, `resolved #1234`

## 📝 What does this PR do?

> Summarize your work here.
> if you have any plots/diagrams/screenshots/tables, please attach them here.

Support all-gather overlap with forward.

No-overlap:
image

Overlap:
image

GPU Util and SM efficiency is higher.

And according to trace, it's well overlaped:
image

## 💥 Checklist before requesting a review

- [ ] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [ ] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [ ] I have performed a self-review of my code
- [ ] I have added thorough tests.
- [ ] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [ ] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

File: colossalai/booster/plugin/hybrid_parallel_plugin.py [hunk 1]
```diff
@@ -879,6 +879,10 @@ def _compute_grad_norm(self, dp_pg, gradients: List[Tensor], norm_type: int = 2)
 
         return total_norm
 
+    def step(self, closure=None):
+        super().step(closure)
+        self._force_waite_all_gather()
```
Comment (FrankLeeeee) [inline colossalai/booster/plugin/hybrid_parallel_plugin.py:884]: `waite` is mis-spelled.

File: examples/language/performance_evaluator.py [hunk 2]
```diff
@@ -113,13 +113,13 @@ def on_step_start(self, step: int) -> None:
         self.disable = self.ignore_steps > 0 and step < self.ignore_steps
         if self.disable:
             return
-        get_accelerator().synchronize()
+        # get_accelerator().synchronize()
```
Comment (FrankLeeeee) [inline examples/language/performance_evaluator.py:116]: Why comment this line?

File: colossalai/zero/low_level/low_level_optim.py [hunk 3]
```diff
@@ -892,3 +900,7 @@ def get_working_grad_by_param_id(self, param_id: int) -> Tensor:
     def get_partitioned_gradients_by_param_id(self, group_id: int, param_id: int) -> List:
         grad_store = self.pid_to_grad_store[param_id]
         return grad_store.get_partitioned_gradients_by_param_id(group_id, param_id)
+
+    def _force_waite_all_gather(self):
```
Comment (FrankLeeeee) [inline colossalai/zero/low_level/low_level_optim.py:904]: What is the diff between `waite` and `wait`?

File: examples/language/performance_evaluator.py (hunk 2 above)
Comment (ver217) [reply examples/language/performance_evaluator.py:116]: There is async communication between `optimizer.step()` and `model(**data)`. Synchronization may make it ineffective.

File: colossalai/zero/low_level/low_level_optim.py (hunk 3 above)
Comment (ver217) [reply colossalai/zero/low_level/low_level_optim.py:904]: typo

**Review by FrankLeeeee — APPROVED** [review]

**Pull request merged**